### PR TITLE
sa,unixsock: add unix domain socket support

### PR DESCRIPTION
**Pull request by sreimers**
Usage example:

```c
	struct sa srv;
	re_sock_t fd;

	err = sa_set_str(&srv, "unix:/run/http.sock", 0);
	if (err)
		return err;

	err = unixsock_listen_fd(&fd, &srv);
	if (err)
		return err;

	re_printf("listen: %j\n", &srv);

	err = http_listen_fd(sock, fd, http_req_handler, mix);
```

### CODE REVIEW

File: include/re_sa.h [hunk 1]
```diff
@@ -6,10 +6,18 @@
 #if defined(WIN32)
 #include <winsock2.h>
 #include <ws2tcpip.h>
+#if !defined(UNIX_PATH_MAX)
+#define UNIX_PATH_MAX 108
+typedef struct sockaddr_un {
+	ADDRESS_FAMILY sun_family;
+	char sun_path[UNIX_PATH_MAX];
+} SOCKADDR_UN, *PSOCKADDR_UN;
+#endif
```
Comment (alfredh) [inline include/re_sa.h:15]: does unixsock work on windows ?

File: include/re_http.h [hunk 2]
```diff
@@ -182,6 +182,8 @@ struct http_conn;
 typedef void (http_req_h)(struct http_conn *conn, const struct http_msg *msg,
 			  void *arg);
 
+int http_listen_fd(struct http_sock **sockp, re_sock_t fd, http_req_h *reqh,
+		   void *arg);
```
Comment (alfredh) [inline include/re_http.h:186]: is re_sock_t used in the api ?

should we consider moving the definition to re_types.h ?

File: src/unixsock/unixsock.c [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+#include <re_dbg.h>
+
+#ifdef WIN32
+#define close closesocket
+#define unlink _unlink
+#endif
+
+
+int unixsock_listen_fd(re_sock_t *fd, const struct sa *local)
+{
```
Comment (alfredh) [inline src/unixsock/unixsock.c:21]: check input arguments

File: include/re_sa.h (hunk 1 above)
Comment (sreimers) [reply include/re_sa.h:15]: Testing is on my list, but its available (including afunix.h does not work for mingw and curl does the same fallback):

https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/

**Comment by alfredh** [comment]
perhaps also update the list of modules:

https://github.com/baresip/re#modules

File: include/re_sa.h [hunk 4]
```diff
@@ -27,6 +35,7 @@ struct sa {
 	union {
 		struct sockaddr sa;
 		struct sockaddr_in in;
+		struct sockaddr_un un;
```
Comment (sreimers) [inline include/re_sa.h:38]: I wonder if we should make this compile optional? like "HAVE_UNIXSOCK", it increases `struct sa` size by 108 bytes.

File: include/re_http.h (hunk 2 above)
Comment (sreimers) [reply include/re_http.h:186]: Done: https://github.com/baresip/re/pull/600/commits/ee74f20808435e6d122a65b9a0c98bfcf258ae7d

File: include/re_types.h [hunk 5]
```diff
@@ -258,3 +258,14 @@ typedef SSIZE_T ssize_t;
...
+/* Socket helpers */
+#ifdef WIN32
+#define ERRNO_SOCK WSAGetLastError()
+#define BAD_SOCK INVALID_SOCKET
+typedef size_t re_sock_t;
+#else
+#define ERRNO_SOCK errno
+#define BAD_SOCK -1
+typedef int re_sock_t;
+#endif
```
Comment (alfredh) [inline include/re_types.h:271]: can you please make this hunk a separate PR ?

I also need it for UDP.

Please also add a `RE_` prefix to ERRNO_SOCK and BAD_SOCK

File: include/re_sa.h (hunk 4 above)
Comment (alfredh) [reply include/re_sa.h:38]: I would prefer positive logic here, i.e. ifdef HAVE_UNIXSOCK

also please note that if libre and application is compiled with different flags,
the size of struct sa would be different, and lead to crash.

File: include/re_unixsock.h [hunk 6]
```diff
@@ -0,0 +1,2 @@
+
+int unixsock_listen_fd(re_sock_t *fd, const struct sa *local);
```
Comment (alfredh) [inline include/re_unixsock.h:2]: perhaps avoid using keyword "local"... maybe "laddr" is better

File: src/sa/sa.c [hunk 7]
```diff
@@ -108,6 +108,11 @@ int sa_pton(const char *addr, struct sa *sa)
 	if (inet_pton(AF_INET, addr, &sa->u.in.sin_addr) > 0) {
 		sa->u.in.sin_family = AF_INET;
 	}
+	else if (!strncmp(addr, "unix:", 5)) {
```
Comment (alfredh) [inline src/sa/sa.c:111]: what if the string is 5 bytes ?

File: src/unixsock/unixsock.c [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	if (listen(*fd, SOMAXCONN) < 0) {
+		err = RE_ERRNO_SOCK;
+		DEBUG_WARNING("listen(): %m (%J)\n", err, local);
+		goto out;
+	}
+
+out:
+	if (err) {
```
Comment (alfredh) [inline src/unixsock/unixsock.c:57]: in error cases the input argument should not be set

File: src/http/server.c [hunk 9]
```diff
@@ -241,6 +241,46 @@ static void connect_handler(const struct sa *peer, void *arg)
...
+ *
+ * @param sockp Pointer to returned HTTP Socket
+ * @param fd    File descriptor
+ * @param reqh  Request handler
+ * @param arg   Handler argument
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int http_listen_fd(struct http_sock **sockp, re_sock_t fd, http_req_h *reqh,
+		   void *arg)
```
Comment (alfredh) [inline src/http/server.c:255]: if you want to make the diff smaller, you can submit this function as a separate PR ...

File: src/sa/sa.c (hunk 7 above)
Comment (sreimers) [reply src/sa/sa.c:111]: Since the provided `addr` string should be always zero terminated, sun_path is empty in this case too.

**Comment by alfredh** [comment]
I think this looks good now.

Please go ahead and merge, if you want
